datafusion/core/src/datasource/schema_adapter.rs

```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
     ) -> datafusion_common::Result<RecordBatch>;
 }
 
-/// Implementation of [`SchemaAdapterFactory`] that maps columns by name
-/// and casts columns to the expected type.
+/// Default  [`SchemaAdapterFactory`] for mapping schemas.
+///
+/// This can be used to adapt file-level record batches to a table schema and
+/// implement schema evolution.
+///
+/// Given an input file schema and a table schema, this factor can make
+/// [`SchemaMapper`]s that:
+///
+/// 1. Reorder columns
+/// 2. Cast columns to the correct type
+/// 3. Fill missing columns with nulls
```

Comment:
What happens if the field is not nullable? I've looked through the code but couldn't find an explicit check for this, so does this error or does this produce invalid record batches? :thinking:

Reply:
It errors -- I will add a test / improve the docs

Reply:
In [b065efe](https://github.com/apache/datafusion/pull/13063/commits/b065efe0456ed88bb7c89aff293acd71eef9cef4)